File: nptel/4_assignment_algorithms.py
import sys
import math
from collections import deque
sys.stdin = open("input.txt", 'r')

source, destination = map(int, input().split())
n, m =  map(int, input().split())
flash_time = list(map(int, input().split()))
flash_time.insert(0,0)

edge_info = []
for i in range(m):
    t = tuple(map(int, input().split()))
    edge_info.append(t)


# Making the graph
graph = {}
for e in edge_info:
    if e[0] in graph:
        graph[e[0]].append((e[1], e[2]))
    else:
        graph[e[0]] = [(e[1], e[2])]

    if e[1] in graph:
        graph[e[1]].append((e[0], e[2]))
    else:
        graph[e[1]] = [(e[0], e[2])]

#print(graph)



visited = {source}
dist = [999999]*(n+1)
dist[source] = 0
current = source
queue = deque()
queue.append(source)

def signal_time(till_now, interval):
    ans = (interval - till_now % interval) % interval
    #return ans if ans != interval else 0
    return ans

while len(visited) < n:
    if dist[current] == 999999:
        break
    for ver in graph[current]:
        if ver[0] not in visited:
            till_now =  dist[current]+ ver[1]
            if ver[0] != destination:
                dist[ver[0]] = min(dist[ver[0]], till_now + signal_time(till_now, flash_time[ver[0]]))
            else:
                dist[ver[0]] = min(dist[ver[0]], till_now)


            #if ver[0] != destination:
            #    till_now = flash_time[ver[0]] * ((till_now + flash_time[ver[0]] - 1) / flash_time[ver[0]])
            #dist[ver[0]] = min(dist[ver[0]], till_now)

            #if till_now < dist[ver[0]]:
            #    dist[ver[0]] = till_now


            queue.append(ver[0])


    print("Visited",current, "weight", dist[current])
    visited.add(current)
    current = queue.popleft()

print(dist[destination], end="")

File: nptel/4_assignment_algorithms.cpp
#include<bits/stdc++.h>
using namespace std;
const int maxNodes=400;
const int infinity=1e9;
vector < pair <int,int> > graph[maxNodes];
int T[maxNodes],dist[maxNodes],vis[maxNodes];
 
int solve(int source,int destination,int n)
{
	set < pair <int,int> > queue;
	for(int i=1;i<=n;++i)
	{
		if(i!=source) {
            dist[i]=infinity;
        }
		queue.insert({dist[i],i});
	}
 
	while(!queue.empty())
	{
		auto top=*queue.begin();
		queue.erase(queue.begin());
		int wt=top.first,curr=top.second;
		vis[curr]=1;
 
		if(wt==infinity) 
            break;
 
		for(auto p:graph[curr]) if(!vis[p.second]) 
		{
			int v=p.second;
			int d=wt+p.first;
			if(v!=destination) {
                d=T[v]*((d+T[v]-1)/T[v]);
            }

			if(d<dist[v])
			{
				queue.erase(queue.find({dist[v],v}));
				dist[v]=d;
				queue.insert({dist[v],v});
			}
		}
	}
	return dist[destination];
}
 
int main()
{
	int n,m,source,destination;
	cin>>source>>destination;
	cin>>n>>m;
 
	for(int i=1;i<=n;++i) {
        cin>>T[i];
    }
 
	for(int p=1;p<=m;++p)
	{
		int i,j,weight;
		cin>>i>>j>>weight;
		graph[i].push_back({weight,j});
		graph[j].push_back({weight,i});
	}
	
    int ans = solve(source,destination,n);
	cout<<ans;
	return 0;
}